### DISCUSSION TITLE
Keycloak with Next-Auth

### DISCUSSION THREAD

**Discussion by ggisbert87** [nextauthjs/next-auth#1195] [Help] [answered]
Hi all, is there any possibility to use Next-Auth with Keycloak as provider?

**Comment by balazsorban44** [comment]
Hi, it is probably possible as I can see it SUPPORTS the OAuth 2/OIDC specs. check this out:

https://next-auth.js.org/configuration/providers#using-a-custom-provider

**Comment by manuschillerdev** [comment, marked as answer]
I am also trying to get keycloak and nextjs up and running. 
Still trying to figure out the right config in Keycloak (https://www.keycloak.org/docs/4.8/authorization_services/#_service_overview) and a custom provider with next-auth. I'll try to post an update here if I make some progress!
Any hints in the right direction are much appreciated! :)

**Reply by manuschillerdev** [reply]
I got it to work:

```js

const options = {
  providers: [
    {
      id: "keycloak",
      name: "Keycloak",
      type: "oauth",
      version: "2.0",
      params: { grant_type: "authorization_code" },
      scope: "openid",
      accessTokenUrl: `${process.env.KEYCLOAK_BASE_URL}/token`,
      requestTokenUrl: `${process.env.KEYCLOAK_BASE_URL}/auth`,
      authorizationUrl: `${process.env.KEYCLOAK_BASE_URL}/auth?response_type=code`,
      clientId: "nextjs_local",
      clientSecret: process.env.KEYCLOAK_CLIENT_SECRET,
      profileUrl: `${process.env.KEYCLOAK_BASE_URL}/userinfo`,
      profile: (profile) => ({ ...profile, id: profile.sub }),
    },
  ],
};
```

```
# env.local
KEYCLOAK_BASE_URL=http://localhost:8080/auth/realms/master/protocol/openid-connect
KEYCLOAK_CLIENT_SECRET=***
NEXTAUTH_URL=http://localhost:3000
```
let me know if you have trouble configuring keycloak :)

**Reply by ggisbert87** [reply]
Thank you, We will try it out!

**Reply by balazsorban44** [reply]
Awesome @manuschillerdev! Would you be interested in adding it to the repository as a dedicated provider?

**Reply by manuschillerdev** [reply]
sure! :) my week is pretty full - if you can wait until the weekend I'll happily contribute :)

**Reply by balazsorban44** [reply]
no pressure at all! do it whenever you have time. 🤗

**Reply by iaincollins** [reply]
Oh this is great to see! I got asked a lot about Keycloak with version 1.0 and because of issues with how it interact with Passport.js it didn't work for most folks back then (though I think a few folks manage to wrangle something, possibly with a fork). Really happy to hear it works much more nicely now, would be a great provider to add if anyone would like to have a go at a PR for this!

**Reply by manuschillerdev** [reply]
first draft for discussion:
https://github.com/nextauthjs/next-auth/issues/1261

once the API question (possible config options) are solved the PR will be pretty quick and easy :)
